Добавлена работа с сетью (фильмы для квиза загружаются из открытого A…

desikvasya opened the pull request:
-В приложении использованы фильмы из топ-250 и самых популярных фильмов IMDb. 
-Обновлён метод viewDidLoad() в MovieQuizViewController 
-Приложение запускается и работает корректно по функциональным и техническим требованиям. 
-Если происходит ошибка загрузки, то показывается алерт с ошибкой. 
-Приложение не пишет в логи сообщения об ошибках, связанных с работой с UI из другого потока.

desikvasya committed:
Добавлена работа с сетью (фильмы для квиза загружаются из открытого API IMDb).
В приложении использованы фильмы из топ-250 и самых популярных фильмов IMDb.
Обновлён метод viewDidLoad() в MovieQuizViewController
Приложение запускается и работает корректно по функциональным и техническим требованиям.
Если происходит ошибка загрузки, то показывается алерт с ошибкой.
Приложение не пишет в логи сообщения об ошибках, связанных с работой с UI из другого потока.

desikvasya committed:
Исправлено после код ревью:
- Нет блокировки кнопок "Да"/"Нет"
- Форматирование кода
- Рамка используется через UIView.animate
- yesButtonClicked, noButtonClicked  теперь сравнивает ответ  currentQuestion.correctAnswer == true /  currentQuestion.correctAnswer == false
-  urlString.components(separatedBy: "._")[0] + "._V0_UX600_.jpg" получение части урла обернул сделал через  [safe: 0]

Doomsday46 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -81,25 +136,37 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+            correctAnswers += 1
+        }
+        
         DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self]  in
```
Какая страшная конструкция), нам не нужен тут `DispatchQueue.main.asyncAfter`. Тебе достаточно оставить UIView.animate.

Doomsday46 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -81,25 +136,37 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
             guard let self = self else {return}
-            self.showNextQuestionOrResults()
-            self.imageView.layer.borderColor = UIColor.clear.cgColor
+            UIView.animate(
```
```
class func animate(
    withDuration duration: [TimeInterval](https://developer.apple.com/documentation/foundation/timeinterval),
    delay: [TimeInterval](https://developer.apple.com/documentation/foundation/timeinterval),
    options: [UIView](https://developer.apple.com/documentation/uikit/uiview).[AnimationOptions](https://developer.apple.com/documentation/uikit/uiview/animationoptions) = [],
    animations: @escaping () -> [Void](https://developer.apple.com/documentation/swift/void),
    completion: (([Bool](https://developer.apple.com/documentation/swift/bool)) -> [Void](https://developer.apple.com/documentation/swift/void))? = nil
)
```

Используй эту сигнатуру метода, если тебе нужен delay перед показом результата. Обрати внимание, что некоторые параметры опциональные и их не нужно проставлять. Так же лучше кнопки разблокировать после отрисовки анимации, так как все равно получается проклинать несколько вопросов сразу

desikvasya replied:
Убрал DispatchQueue, и блокировку кнопок поднял выше к completion: 
Сейчас правильно ? (я пробовал, теперь несколько вопросов нельзя прокликать)

UIView.animate(
              withDuration: 1.0,
              animations: {
                self.imageView.layer.borderColor = UIColor.clear.cgColor
              }, completion: { [weak self] _ in
                self?.showNextQuestionOrResults()
                self?.yesButton.isEnabled = true
                self?.noButton.isEnabled = true
              }
            )

Doomsday46 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -9,74 +9,154 @@ import Foundation
 class QuestionFactory: QuestionFactoryProtocol {
-    func requestNextQuestion() -> QuizQuestion? {
-        return nil
+    func reloadImage() {
```
Пустой метод?) Если так получается, что тебе не нужны методы в классе от протокола, то лучше разделить протоколы и использовать только нужные. Но тут по кнопке reload не вижу кода для перезагрузки контента и показа pop-up c ошибкой

Doomsday46 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,6 +1,12 @@
 import UIKit
 
 final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    func didFailToLoadImage(with error: Error) {
```
Методы не должны описывать в классе до свойств

desikvasya commented:
@Doomsday46 у меня такой вопрос, я заблочил кнопки таким образом, перенес их активность сюда: 
private func show(quiz step: QuizStepViewModel) {
        imageView.image = step.image
        textLabel.text = step.question
        counterLabel.text = step.questionNumber
        yesButton.isEnabled = true
        noButton.isEnabled = true
    }
    Если я делаю так, то проблемы с белой подложной и кнопкой релоад в принципе не возникает, так как если интернет пропадает между вопросами, то следующий вопрос подгружается сам после того как интернет снова есть. Пока интернета нет, кнопки неактивны, и висит картинка прошлого вопроса и счетчик не меняется.
    
    ЛИБО 
    
    я все-таки сделал эту кнопку, и это работает как-то так https://youtube.com/shorts/17PN5zypghs
    
    Пока что у меня с этой кнопкой больше костылей и мучений, другие ребята с таким замечанием не сталкивались, а у наставников нет более менее однозначной рекомендации как это лучше сделать. 
    
    Чтобы работа была принята по какому пути мне лучше пойти ?

desikvasya committed:
- Добавил кнопку релоад если между вопросами нет интернета
- Заблокировал кнопки да/нет если между вопросами нет интернета

desikvasya commented:
@Doomsday46 проверь пожалуйста
https://github.com/desikvasya/Sprint5/tree/Sprint6_Done

desikvasya committed:
Приложение работает и загружает вопросы и картинки из интернета
Исправил следующие моменты после первого код-ревью:
-блокировка кнопок вопроса (больше 10 раз не пройти)
-исправил форматирование во всех файлах через ctrl+i (также убрал лишние функции)
- в showAnswerResult использовал UIVIEW Animate
- yesButtonClicked, noButtonClicked лучше не заводить bool переменную с конкретным значением, а сразу сравнить текущий ответ currentQuestion.correctAnswer == true /  currentQuestion.correctAnswer == false
- обернул получение картинки в проверку
- кнопку убрал, все что от нее осталось закомментил, с куратором поговорил об этом, тяжело далась эта кнопка, такое чувство будто мы еще это не проходили. У наставников тоже не было однозначного ответа как ее лучше реализовать
